### PR TITLE
add sparse update momentum. test=develop

### PR DESCRIPTION
**Pull request by dzhwinter**
fix sparse update in selected rows

### CODE REVIEW

File: paddle/fluid/operators/momentum_op.cc [hunk 1]
```diff
@@ -110,6 +127,8 @@ else:   \\
 }  // namespace paddle
 
 namespace ops = paddle::operators;
-REGISTER_OP_WITHOUT_GRADIENT(momentum, ops::MomentumOp, ops::MomentumOpMaker);
+REGISTER_OPERATOR(momentum, ops::MomentumOp, ops::MomentumOpMaker,
```
Comment (jacquesqiao) [inline paddle/fluid/operators/momentum_op.cc:142]: momentum_op is optimizer, has no gradient operator.

Comment (dzhwinter) [reply paddle/fluid/operators/momentum_op.cc:142]: I agree, so I use EmptyGradOpMaker. Switch from REGISTER_OP_WITHOUT_GRADIENT to REGISTER_OPERATOR is to support OpInferVarType.

Comment (panyx0718) [reply paddle/fluid/operators/momentum_op.cc:142]: Perhaps REGISTER_OP_WITHOUT_GRADIENT should support OpInferVarType

File: paddle/fluid/operators/momentum_op.cc [hunk 2]
```diff
@@ -53,13 +53,30 @@ class MomentumOp : public framework::OperatorWithKernel {
...
+ public:
+  void operator()(const framework::OpDesc& op_desc,
+                  framework::BlockDesc* block) const override {
+    auto input_var = op_desc.Input("Param")[0];
+    for (auto& out_var : op_desc.Output("ParamOut")) {
+      if (block->FindRecursiveOrCreateVar(input_var).GetType() ==
+          framework::proto::VarType::SELECTED_ROWS) {
+        block->FindRecursiveOrCreateVar(out_var).SetType(
+            framework::proto::VarType::SELECTED_ROWS);
+      } else {
```
Comment (panyx0718) [inline paddle/fluid/operators/momentum_op.cc]: else if? I'm afraid there are many different types.
else throw exception, say no supported

**Review by panyx0718 — DISMISSED** [review]

File: paddle/fluid/operators/momentum_op.cu [hunk 3]
```diff
@@ -42,32 +42,92 @@ __global__ void MomentumKernel(const T* p, const T* g, const T* v,
...
-    MomentumKernel<T><<<grid, block, 0, ctx.cuda_device_context().stream()>>>(
-        p, g, v, lr, mu, param->numel(), use_nesterov, p_out, v_out);
+      const int kThreadPerBlock = 256;
+      int grid = (param->numel() + kThreadPerBlock - 1) / kThreadPerBlock;
+      SparseMomentumKernel<
+          T><<<grid, kThreadPerBlock, 0, ctx.cuda_device_context().stream()>>>(
+          p, g, v, lr, mu, rows.CUDAData(ctx.GetPlace()), grad_row_numel,
+          grad->rows().size(), use_nesterov, p_out, v_out);
+    } else {
+      PADDLE_THROW("Unsupported Variable Type of Grad");
```
Comment (chengduoZH) [inline paddle/fluid/operators/momentum_op.cu]: It is better that the error should contain more detail information, e.g.
```
PADDLE_THROW("MomentumOp only supports LoDTensor or SelectedRows gradient, but the received Variable Type is %s", grad_var->Type().name());
```

File: paddle/fluid/operators/momentum_op.cu [hunk 4]
```diff
@@ -42,32 +42,92 @@ __global__ void MomentumKernel(const T* p, const T* g, const T* v,
...
+                                     const T* lr, const T mu,
+                                     const int64_t* grad_rows,
+                                     const size_t grad_row_numel,
+                                     const size_t grad_row_size,
+                                     const T use_nesterov, T* p_out, T* v_out) {
+  for (int i = blockIdx.x; i < grad_row_size; i += gridDim.x) {
+    for (int j = threadIdx.x; j < grad_row_numel; j += blockDim.x) {
+      size_t p_i = grad_rows[i] * grad_row_numel + j;
+      size_t g_i = i * grad_row_numel + j;
+      v_out[g_i] = v[g_i] * mu + g[g_i];
```
Comment (chengduoZH) [inline paddle/fluid/operators/momentum_op.cu]: According to mathematical equation of `Momentum`, all the velocity should be updated.

File: paddle/fluid/operators/momentum_op.cc (hunk 2 above)
Comment (dzhwinter) [reply paddle/fluid/operators/momentum_op.cc]: Yes.  Given that this one depends on https://github.com/PaddlePaddle/Paddle/pull/13689/files. 
Another PR will be sent to refine here.

File: paddle/fluid/operators/momentum_op.cu (hunk 4 above)
Comment (dzhwinter) [reply paddle/fluid/operators/momentum_op.cu]: fixed.

File: paddle/fluid/operators/momentum_op.h [hunk 5]
```diff
@@ -59,5 +71,282 @@ class MomentumOpKernel : public framework::OpKernel<T> {
...
+      merge_func(ctx.template device_context<DeviceContext>(), *grad,
+                 merged_grad);
+
+      platform::ForRange<DeviceContext> for_range(
+          static_cast<const DeviceContext&>(ctx.device_context()),
+          param->numel());
+
+      const int64_t* rows = nullptr;
+      if (platform::is_gpu_place(ctx.GetPlace())) {
+        rows = merged_grad->rows().CUDAData(ctx.GetPlace());
```
Comment (sneaxiy) [inline paddle/fluid/operators/momentum_op.h:314]: Maybe use `#ifdef PADDLE_WITH_CUDA` here? I remember the CPU version of `mixed_vector` is just an alias of `std::vector`, which has no `CUDAData` method.

Comment (dzhwinter) [reply paddle/fluid/operators/momentum_op.h:314]: fixed.

**Review by sneaxiy — APPROVED** [review]
Excellent!

**Pull request merged**